File: packages/core/src/storage.ts
/**
 * Storage abstraction — PROTOCOL.md §8.
 *
 * The engine talks ONLY to this interface. Adapters are dumb byte movers with
 * one hard obligation: applyBatch is atomic (§8.2). MemoryAdapter is the
 * reference used by tests; IndexedDB/SQLite adapters come in M2+.
 */

import type { Frontier, Op } from "./op.js";

export interface ClockState {
  readonly millis: number;
  readonly counter: number;
}

export interface BatchWrite {
  /** ops to append to the log (already deduped by the engine) */
  readonly ops: readonly Op[];
  /** new winning op per affected cell, keyed by cellKey() */
  readonly winners: ReadonlyMap<string, Op>;
  /** frontier after this batch */
  readonly frontier: Frontier;
  /** clock state after this batch */
  readonly clock: ClockState;
}

/** Ops and cells to physically remove during compaction (§9). */
export interface CompactionWrite {
  /** op ids to delete from the log */
  readonly opIds: readonly string[];
  /** materialized cell keys to forget (tombstone GC only) */
  readonly cellKeys: readonly string[];
}

export interface StorageAdapter {
  /** All winning cells of a row: column → winning op. undefined if none. */
  getRow(table: string, row: string): Promise<ReadonlyMap<string, Op> | undefined>;
  /** rowIds that have at least one cell op in this table (incl. tombstoned). */
  listRows(table: string): Promise<string[]>;
  /** Table names that have at least one op. */
  listTables(): Promise<string[]>;
  hasOp(id: string): Promise<boolean>;
  /** Current winner for one cell. */
  getWinner(table: string, row: string, column: string): Promise<Op | undefined>;
  /** Every stored op strictly above `frontier`, sorted by hlc. §6 step 3. */
  opsSince(frontier: Frontier): Promise<Op[]>;
  getFrontier(): Promise<Frontier>;
  getClock(): Promise<ClockState | undefined>;
  /** Atomic, all-or-nothing. §8.2. */
  applyBatch(batch: BatchWrite): Promise<void>;

  // ---- compaction support (§9) ----

  /** Total ops currently retained in the log. */
  opCount(): Promise<number>;
  /** Every op in the log, ascending by hlc. Used by compaction scans. */
  allOps(): Promise<Op[]>;
  /** Last known frontier of each peer, from since/ack exchanges (§6). */
  getPeerFrontiers(): Promise<Record<string, Frontier>>;
  /** Record a peer's frontier. */
  setPeerFrontier(peer: string, frontier: Frontier): Promise<void>;
  /** Atomically remove ops (and, for tombstone GC, cells). */
  compact(write: CompactionWrite): Promise<void>;
}

const SEP = "\u0000";
export function cellKey(table: string, row: string, column: string): string {
  return table + SEP + row + SEP + column;
}

export class MemoryAdapter implements StorageAdapter {
  private opsById = new Map<string, Op>();
  /** table → row → column → winning op */
  private tables = new Map<string, Map<string, Map<string, Op>>>();
  private frontier: Frontier = {};
  private clock: ClockState | undefined;
  private peerFrontiers: Record<string, Frontier> = {};

  async getRow(table: string, row: string): Promise<ReadonlyMap<string, Op> | undefined> {
    return this.tables.get(table)?.get(row);
  }

  async listRows(table: string): Promise<string[]> {
    return [...(this.tables.get(table)?.keys() ?? [])];
  }

  async listTables(): Promise<string[]> {
    return [...this.tables.keys()];
  }

  async hasOp(id: string): Promise<boolean> {
    return this.opsById.has(id);
  }

  async getWinner(table: string, row: string, column: string): Promise<Op | undefined> {
    return this.tables.get(table)?.get(row)?.get(column);
  }

  async opsSince(frontier: Frontier): Promise<Op[]> {
    const out: Op[] = [];
    for (const op of this.opsById.values()) {
      const seen = frontier[op.device];
      if (seen === undefined || op.hlc > seen) out.push(op);
    }
    out.sort((a, b) => (a.hlc < b.hlc ? -1 : a.hlc > b.hlc ? 1 : 0));
    return out;
  }

  async getFrontier(): Promise<Frontier> {
    return this.frontier;
  }

  async getClock(): Promise<ClockState | undefined> {
    return this.clock;
  }

  async opCount(): Promise<number> {
    return this.opsById.size;
  }

  async allOps(): Promise<Op[]> {
    return [...this.opsById.values()].sort((a, b) =>
      a.hlc < b.hlc ? -1 : a.hlc > b.hlc ? 1 : 0,
    );
  }

  async getPeerFrontiers(): Promise<Record<string, Frontier>> {
    return this.peerFrontiers;
  }

  async setPeerFrontier(peer: string, frontier: Frontier): Promise<void> {
    this.peerFrontiers = { ...this.peerFrontiers, [peer]: frontier };
  }

  async compact(write: CompactionWrite): Promise<void> {
    const newOps = new Map(this.opsById);
    for (const id of write.opIds) newOps.delete(id);
    const newTables = new Map(this.tables);
    for (const key of write.cellKeys) {
      const [table, row, column] = key.split(SEP) as [string, string, string];
      const rows = new Map(newTables.get(table) ?? []);
      const cells = new Map(rows.get(row) ?? []);
      cells.delete(column);
      if (cells.size === 0) rows.delete(row);
      else rows.set(row, cells);
      if (rows.size === 0) newTables.delete(table);
      else newTables.set(table, rows);
    }
    this.opsById = newOps;
    this.tables = newTables;
  }

  async applyBatch(batch: BatchWrite): Promise<void> {
    // In-memory "transaction": build everything, then swap. A thrown error
    // before the swap leaves prior state untouched.
    const newOps = new Map(this.opsById);
    for (const op of batch.ops) newOps.set(op.id, op);
    const newTables = new Map(this.tables);
    for (const [key, op] of batch.winners) {
      const [table, row, column] = key.split(SEP) as [string, string, string];
      const rows = new Map(newTables.get(table) ?? []);
      const cells = new Map(rows.get(row) ?? []);
      cells.set(column, op);
      rows.set(row, cells);
      newTables.set(table, rows);
    }
    this.opsById = newOps;
    this.tables = newTables;
    this.frontier = batch.frontier;
    this.clock = batch.clock;
  }
}

File: impl/dart/lib/src/storage.dart
/// Storage — PROTOCOL.md section 8.
///
/// The protocol mandates capabilities, not an engine. Implement
/// [StorageAdapter] against whatever your platform offers; the one
/// non-negotiable is that [applyBatch] is atomic (section 8.2).
library;

import 'hlc.dart';
import 'signing.dart';
import 'op.dart';

/// One atomic unit of work: log append, cell updates and frontier advance.
///
/// These MUST land together. A crash between the log write and the cell write
/// leaves a replica that disagrees with itself and converges wrongly forever.
class BatchWrite {
  const BatchWrite({
    required this.ops,
    required this.winners,
    required this.frontier,
    this.clock,
  });

  /// New ops to append to the log.
  final List<Op> ops;

  /// Cells whose winning op changed, keyed by (table, row, column).
  final Map<CellKey, Op> winners;

  /// The frontier after applying [ops].
  final Frontier frontier;

  /// Local clock state to persist, when a local write advanced it.
  final Hlc? clock;
}

/// Identity of one cell. Section 8.
class CellKey {
  const CellKey(this.table, this.row, this.column);

  final String table;
  final String row;
  final String column;

  @override
  bool operator ==(Object other) =>
      other is CellKey &&
      other.table == table &&
      other.row == row &&
      other.column == column;

  @override
  int get hashCode => Object.hash(table, row, column);

  @override
  String toString() => '$table/$row/$column';
}

abstract class StorageAdapter {
  /// All winning cells of a row: column to winning op. Null if none.
  Future<Map<String, Op>?> getRow(String table, String row);

  /// Row ids with at least one cell op in this table, including tombstoned.
  Future<List<String>> listRows(String table);

  /// Table names with at least one op.
  Future<List<String>> listTables();

  Future<bool> hasOp(String id);

  Future<Op?> getWinner(String table, String row, String column);

  /// Every stored op strictly above [frontier], ascending by hlc.
  /// Section 6 step 3.
  Future<List<Op>> opsSince(Frontier frontier);

  Future<Frontier> getFrontier();

  Future<Hlc?> getClock();

  /// The device's signing keypair, or null on a fresh store. Section 12.
  ///
  /// Stored in the clear beside the data it protects. On a device this belongs
  /// in Keychain or Keystore; the space secret already carries the same
  /// exposure, so this does not widen it.
  Future<DeviceKey?> getDeviceKey();
  Future<void> setDeviceKey(DeviceKey key);

  /// Atomic, all-or-nothing. Section 8.2.
  Future<void> applyBatch(BatchWrite batch);

  Future<int> opCount();

  /// Every op in the log, ascending by hlc.
  Future<List<Op>> allOps();
}

/// Reference in-memory adapter.
///
/// Not durable, but it is the shortest complete statement of what an adapter
/// has to do, and it is what the conformance vectors run against first.
class MemoryAdapter implements StorageAdapter {
  final Map<String, Op> _log = {};
  final Map<CellKey, Op> _winners = {};
  Frontier _frontier = {};
  Hlc? _clock;
  DeviceKey? _deviceKey;

  @override
  Future<Map<String, Op>?> getRow(String table, String row) async {
    Map<String, Op>? out;
    for (final entry in _winners.entries) {
      if (entry.key.table == table && entry.key.row == row) {
        (out ??= {})[entry.key.column] = entry.value;
      }
    }
    return out;
  }

  @override
  Future<List<String>> listRows(String table) async {
    final rows = <String>{};
    for (final key in _winners.keys) {
      if (key.table == table) rows.add(key.row);
    }
    final list = rows.toList()..sort();
    return list;
  }

  @override
  Future<List<String>> listTables() async {
    final tables = <String>{for (final key in _winners.keys) key.table};
    final list = tables.toList()..sort();
    return list;
  }

  @override
  Future<bool> hasOp(String id) async => _log.containsKey(id);

  @override
  Future<Op?> getWinner(String table, String row, String column) async =>
      _winners[CellKey(table, row, column)];

  @override
  Future<List<Op>> opsSince(Frontier frontier) async {
    final out = _log.values.where((op) => aboveFrontier(op, frontier)).toList();
    out.sort((a, b) => a.hlc.compareTo(b.hlc));
    return out;
  }

  @override
  Future<Frontier> getFrontier() async => Map.of(_frontier);

  @override
  Future<DeviceKey?> getDeviceKey() async => _deviceKey;

  @override
  Future<void> setDeviceKey(DeviceKey key) async => _deviceKey = key;

  @override
  Future<Hlc?> getClock() async => _clock;

  @override
  Future<void> applyBatch(BatchWrite batch) async {
    // Single-threaded and synchronous, so this is trivially atomic. A durable
    // adapter needs a real transaction here.
    for (final op in batch.ops) {
      _log[op.id] = op;
    }
    _winners.addAll(batch.winners);
    _frontier = Map.of(batch.frontier);
    if (batch.clock != null) _clock = batch.clock;
  }

  @override
  Future<int> opCount() async => _log.length;

  @override
  Future<List<Op>> allOps() async {
    final out = _log.values.toList();
    out.sort((a, b) => a.hlc.compareTo(b.hlc));
    return out;
  }
}
